[qat] ConvBn2d/ConvBnReLU2d

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23357 [qat] ConvBn2d/ConvBnReLU2d**

Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
[qat] ConvBn2d/ConvBnReLU2d

Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+        self.running_var.fill_(1)
+        self.num_batches_tracked.zero_()
+
+    def reset_parameters(self):
```
If you inherit from Conv2d, you could do super().reset_parameters() and then do the rest of the reset.

jerryzh168 replied:
`reset_parameters()` is  called in `__init__` of `Conv2d`, so if we inherit from `Conv2d`, we'll call `reset_parameters()` of `ConvBnReLU2d` which contains some member variables that are not initialized yet. That's why I'm inheriting from `_ConvNdBase` to avoid the call of `reset_parameters` in `__init__`

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+        return self
+
+    def conv2d_forward(self, input, weight):
+        if self.padding_mode == 'circular':
```
Didnt you share this code with Conv2d and call this via super()?

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+                                       groups, bias, padding_mode)
+        self.eps = eps
+        self.momentum = momentum
+        self.freeze_bn = freeze_bn
```
if self.training:
  self.freeze_bn = freeze_bn
else:
 self.freeze_bn = False

zafartahirov replied:
Or a one-liner:

```python
self.freeze_bn = not self.training or freeze_bn
```

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+import torch.nn.functional as F
+
+
+class ConvBn2d(_ConvNdBase):
```
Similar comment: Might be  better to inherit from nn.Conv2d

jerryzh168 replied:
see previous comment

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import math
+import torch
+from torch.nn.modules.conv import _ConvNdBase
```
from torch.nn.modules.conv import _ConvNdBase: Gives an import error, did you mean _ConvNd? Why not Conv2d?

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_relu.py:
```diff
@@ -1,7 +1,6 @@
-from torch.nn._intrinsic import ConvReLU2d as NNConvReLU2d
-from torch.quantization.QConfig import default_qat_qconfig
+from ....qat.modules.conv import Conv2d as QATConv2d
+from ...modules.fused import ConvReLU2d as NNConvReLU2d
```
NIT: Why have conv_bn and conv_bn_relu in one file and conv_relu in another. Will it be better to have them in one file?

jerryzh168 replied:
sure

raghuramank100 commented on torch/nn/modules/conv.py:
```diff
@@ -18,7 +17,7 @@ class _ConvNd(Module):
                  padding, dilation, transposed, output_padding,
                  groups, bias, padding_mode):
-        super(_ConvNd, self).__init__()
+        super(_ConvNdBase, self).__init__()
```
What is the need to create _ConvNdBase? It seems to be identical to _ConvNd but for reset_parameters()? If you inherit from Conv2d, this change will not be needed.

jerryzh168 replied:
see comments above

raghuramank100 reviewed (changes requested):
Please see inline comments, thanks

zafartahirov commented on test/test_qat.py:
```diff
@@ -0,0 +1,126 @@
+
+class IntrinsicQATModuleTest(TestCase):
+
+    @given(batch_size=st.integers(1, 3),
```
FYI: There is a `tensor_conv2d` strategy in the `hypothesis utils

jerryzh168 replied:
I looked at `tensor_conv2d`, why don't you include generation of `pad`, `stride` in it as well? I can migrate when everything is included.

zafartahirov commented on test/test_qat.py:
```diff
@@ -0,0 +1,126 @@
+            use_bias,
+            padding_mode
+        ).to(dtype=torch.float)
+        bn_op = BatchNorm2d(output_channels, eps, momentum).to(dtype=torch.float)
```
Do we test if it is affine or not?

jerryzh168 replied:
we will enforce affine at the time of fusion.

zafartahirov commented on test/test_qat.py:
```diff
@@ -0,0 +1,126 @@
+        bn_op.bias = qat_op.beta
+
+        def compose(functions):
+            return reduce(lambda f, g: lambda x: f(g(x)), functions, lambda x: x)
```
Should it be `*functions`? As in

```python
def compose(*functions):
    # Create a lambda that applies 'functions' starting from the last
    return reduce(lambda f, g: lambda x: f(g(x)), *functions, lambda x: x)
```

jerryzh168 replied:
I'm passing in a list, so it is fine.

zafartahirov commented on test/test_qat.py:
```diff
@@ -0,0 +1,126 @@
+
+        def compose(functions):
+            return reduce(lambda f, g: lambda x: f(g(x)), functions, lambda x: x)
+        func_list = [bn_op, conv_op]
```
We can rewrite it like so, and put the comment in the `compose` that it reverses the rule application

```python
func_list = [conv_op, bn_op]
if use_relu:
  func_list.append(relu_op)
ref_op = compose(func_list[::-1])
```

jerryzh168 replied:
OK

jerryzh168 replied:
I think it's fine, the order of composition if right to left.

jerryzh168 replied:
changed order

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

zafartahirov commented on torch/nn/_intrinsic/qat/modules/conv_bn_relu.py:
```diff
@@ -0,0 +1,227 @@
+        weight_fake_quant: fake quant module for weight
+
+    """
+    __FLOAT_MODULE__ = NNConvBn2d
```
I don't think we should use magic naming here. `_FLOAT_MODULE` is sufficient

jerryzh168 replied:
I'll change to `__FLOAT_MODULE` since it's a class attribute.
https://pep8.org/#descriptive-naming-styles

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+                 # bn args
+                 # num_features: enforce this matches out_channels before fusion
+                 eps=1e-05, momentum=0.1,
+                 # affine: enforce this is True before fusion?
```
Please remove comments

jerryzh168 replied:
I think we want to keep these for documentation purpose?

raghuramank100 replied:
Sure, please remove the comments with question marks like #affine etc

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+        init.uniform_(self.gamma)
+        init.zeros_(self.beta)
+
+    def reset_parameters(self):
```
You could do:
```python
def reset_parameters(self):
        super(ConvBn2d, self).reset_parameters()
        # A hack to avoid resetting on undefined parameters
        if hasattr(self, 'gamma'):
            self.reset_bn_parameters()
```

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+        scaled_weight = self.weight * scale_factor.reshape([-1, 1, 1, 1])
+        conv = self.conv2d_forward(input, self.weight_fake_quant(scaled_weight))
+
+        print('bias:', self.bias)
```
Please remove prints

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+        else:
+            conv_orig = conv / scale_factor.reshape([1, -1, 1, 1])
+        print('conv_orig:', conv_orig)
+        batch_mean = torch.mean(conv_orig, dim=[0, 2, 3])
```
batch_mean and variance should be computed only if self.training is True

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+            conv = conv + (self.beta - self.gamma * self.running_mean /
+                           torch.sqrt(self.running_var + self.eps)).reshape([1, -1, 1, 1])
+
+        self.running_mean = exponential_average_factor * batch_mean + (1 - exponential_average_factor) * self.running_mean
```
running mean and var should be updated only if training is True and freeze_bn is False:
```
if self.training and not self.freeze_bn:
      self.running_mean = exponential_average_factor * batch_mean + (1 - exponential_average_factor) * self.running_mean       
      self.running_var = exponential_average_factor * batch_var + (1 - exponential_average_factor) * self.running_var
```

jerryzh168 replied:
yeah, just did that

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,265 @@
+                 in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1,
+                 bias=True, padding_mode='zeros',
+                 # bn args
```
Please remove comments

raghuramank100 reviewed (changes requested):
Please see comments

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+        if self.training and not self.freeze_bn:
+            # recovering original conv to get original batch_mean and batch_var
+            conv_orig = conv
+            if self.bias is not None:
```
We should assert that bias is None at init. When conv and bn are present, conv should not have a bias.

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+                 # Conv2d args
+                 in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1,
+                 bias=True, padding_mode='zeros',
```
bias should not be an argument

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+                conv_orig = conv_orig - self.bias.reshape([1, -1, 1, 1])
+            conv_orig = conv_orig / scale_factor.reshape([1, -1, 1, 1])
+            batch_mean = torch.mean(conv_orig, dim=[0, 2, 3])
+            if self.bias is not None:
```
Same here, please remove reference to bias being present

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+        qat_convbn.weight = conv.weight
+        qat_convbn.bias = conv.bias
+        qat_convbn.gamma = bn.weight
+        qat_convbn.beta = bn.bias
```
Add an assert check to ensure that conv.bias is False

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+                 # Conv2d args
+                 in_channels, out_channels, kernel_size, stride=1,
+                 padding=0, dilation=1, groups=1,
+                 bias=True, padding_mode='zeros',
```
Similar comments here

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,267 @@
+                 bias=True, padding_mode='zeros',
+                 activation_fake_quant=None,
+                 weight_fake_quant=None):
+        super(ConvReLU2d, self).__init__(in_channels, out_channels, kernel_size,
```
This is the only case where having bias as an option makes sense.

raghuramank100 reviewed (changes requested):
Please remove bias as an option for fused conv bn and conv bn relu.

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,255 @@
+            # recovering original conv to get original batch_mean and batch_var
+            conv_orig = conv / scale_factor.reshape([1, -1, 1, 1])
+            batch_mean = torch.mean(conv_orig, dim=[0, 2, 3])
+            if self.bias is not None:
```
Please remove if self.bias and corresponding code here, self.bias should be None.

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -0,0 +1,255 @@
+                         qconfig.activation,
+                         qconfig.weight)
+
+        qat_convbn.weight = conv.weight
```
Please add assert to check that bias is False for conv here.

raghuramank100 reviewed (approved):
Please address remaining comments.

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

raghuramank100 commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -165,7 +163,7 @@ def from_float(cls, mod, qconfig):
                          qconfig.activation,
                          qconfig.weight)
-
+        assert qat_convbn.bias is None, 'QAT ConvBn should no bias'
```
NIT: assert qat_convbn.bias is None, 'QAT ConvBn: Conv should not have bias'

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)

jerryzh168 committed:
Update on "[qat] ConvBn2d/ConvBnReLU2d"


Added _intrinsic.qat.ConvBn2d/_intrinsic.qat.ConvBnReLU2d.

Differential Revision: [D16295500](https://our.internmc.facebook.com/intern/diff/D16295500/)
